core/src/solution_submission.rs

```diff
@@ -244,36 +272,86 @@ async fn retry_with_gas_price_increase(
-            gas_price_increase_count, gas_price
+            gas_price_increase_count,
+            gas_price
         );
         let is_last_iteration = gas_price >= effective_gas_price_cap;
+        // In the last iteration there is no block timeout. This is important for the deadline
+        // logic because we assume that in the last iteration the solution future will never be
+        // a confirm timeout so that it either completes in a way we don't need to cancel or we hit
+        // the deadline.
         let block_timeout = if is_last_iteration {
             None
         } else {
             Some(BLOCK_TIMEOUT)
         };
-        let result = contract
-            .submit_solution(
-                batch_index,
-                solution.clone(),
-                claimed_objective_value,
-                gas_price,
-                block_timeout,
-            )
-            .await;
-        if !is_confirm_timeout(&result) || is_last_iteration {
-            return result;
+        let solution_future = contract.submit_solution(
+            batch_index,
+            solution.clone(),
+            claimed_objective_value,
+            gas_price,
+            block_timeout,
+        );
+        let deadline_future = wait_for_deadline(deadline);
+        futures::pin_mut!(deadline_future);
+        let select_future = future::select(solution_future, deadline_future);
+        match determine_retry_action(select_future.await, is_last_iteration) {
+            RetryAction::IncreaseGasPrice => (),
+            RetryAction::Cancel => {
+                let gas_price = U256::from(
+                    (gas_price.as_u128() as f64 * MIN_GAS_PRICE_INCREASE_FACTOR).ceil() as u128,
+                );
+                match contract.send_noop_transaction(gas_price).await {
+                    Ok(_) => log::info!(
+                        "cancelled solution submission of batch {} because of deadline",
+                        batch_index
+                    ),
+                    Err(err) => log::error!(
+                        "failed to cancel solution submission of batch {} after deadline: {:?}",
+                        batch_index,
+                        err
+                    ),
+                }
+                return Err(RetryError::TransactionNotConfirmedInTime);
+            }
+            RetryAction::Done(result) => return result,
         }
     }
     unreachable!("increased gas price past expected limit");
 }
 
+#[derive(Debug)]
+enum RetryAction {
+    IncreaseGasPrice,
+    Cancel,
+    Done(std::result::Result<(), RetryError>),
+}
+
+// Either::Left is the solution submission future, Either::Right the deadline future.
+fn determine_retry_action<T0, T1>(
+    either: Either<(Result<(), MethodError>, T0), ((), T1)>,
+    is_last_iteration: bool,
+) -> RetryAction {
+    match either {
+        Either::Left((result, _)) => {
+            if !is_confirm_timeout(&result) || is_last_iteration {
```

Comment:
Do we even need the `|| is_last_iteration` if in the last iteration the confirm will never timeout?

Reply:
Updated this if condition